harness/determined/common/api/authentication.py

```diff
@@ -62,139 +69,79 @@ def default_load_user_password(
+    master_address: Optional[str] = None,
+    requested_user: Optional[str] = None,
+    password: Optional[str] = None,
     cert: Optional[certs.Cert] = None,
-) -> str:
-    password = api.salt_and_hash(password)
-    unauth_session = api.Session(user=username, master=master_address, auth=None, cert=cert)
-    login = bindings.v1LoginRequest(username=username, password=password, isHashed=True)
-    r = bindings.post_Login(session=unauth_session, body=login)
-    token = r.token
-
-    return token
-
-
-class LogoutAuthentication(Authentication):
+) -> UsernameTokenPair:
     """
-    An api-compatible Authentication object that is basically exactly a UserTokenPair.
+    Log in, preferring cached credentials in the TokenStore, if possible.
 
-    TODO(MLG-215): delete Authentication class and write a function that returns a UsernameTokenPair
-    in its place, and let do_request() take UsernameTokenPair as input.
+    This is the login path for nearly all user-facing cases.
     """
 
-    def __init__(self, session_user: str, session_token: str) -> None:
-        self.session_user = session_user
-        self.session_token = session_token
+    master_address = master_address or util.get_default_master_address()
+    token_store = TokenStore(master_address)
 
-    def get_session_user(self) -> str:
-        return self.session_user
+    # Get session_user and password given the following priority:
+    # 1. User passed in with flag (requested_user)
+    # 2. User from environment if DET_PASS is set.
+    # 3. Active user from the token store.
+    user, password = default_load_user_password(requested_user, password, token_store)
+
+    # For login, we allow falling back to the default username.
+    if not user:
+        user = constants.DEFAULT_DETERMINED_USER
+    assert user is not None
+
+    # Check the token store if this session_user has a cached token. If so, check with the
+    # master to verify it has not expired. Otherwise, let the token be None.
+    token = token_store.get_token(user)
+    if token is not None and not _is_token_valid(master_address, token, cert):
+        token_store.drop_user(user)
+        token = None
+
+    if token is not None:
+        return UsernameTokenPair(user, token)
+
+    # Special case: use token provided from the container environment if:
+    # - No token was obtained from the token store already,
+    # - There is a token available from the container environment, and
+    # - No user was explicitly requested, or the requested user matches the token available in the
+    #   container environment.
+    if (
+        util.get_det_username_from_env() is not None
+        and util.get_det_user_token_from_env() is not None
+        and requested_user in (None, util.get_det_username_from_env())
+    ):
+        user = util.get_det_username_from_env()
+        assert user
+        token = util.get_det_user_token_from_env()
+        assert token
+        return UsernameTokenPair(user, token)
```

Comment:
Should this token be added to the token store?

Reply:
No; if you did, you wouldn't be able to distinguish between the "default master-provided login" and a "user explicitly logged in via `det user login` inside the container".

Reply:
Again going back to definitions, I don't know what it means to explicitly log in.

I maybe also don't know what the token store is supposed to do, because I can't yet reason for myself why this token shouldn't be added.

Further explanation might be best, here, on a call.

Reply:
> the "default master-provided login" and a "user explicitly logged in via det user login inside the container".

I think this is what `active_user` is for?

Reply:
Ok, we talked more.

I was incorrect in what I said earlier.  If you added this to the token store, you wouldn't _have_ to distinguish between the two kinds of logins.

Still, I think it's better as-is, because it keeps things separate.  And the runtime behavior is identical in either case, and you need the same if statements.  If you detected the environment variable and put it into the active_store yourself the first time you detect the environment variable, then you'd go through a different code path on the next execution... and that feels needlessly complicated to me

Reply:
Seconded. Thanks for writing out your opinion.